libafl/src/mutators/string.rs

```diff
@@ -0,0 +1,599 @@
+        .map(|idx| haystack[idx])
+}
+
+fn find_range<F: Fn(char) -> bool>(
+    chars: &[(usize, char)],
+    idx: usize,
+    predicate: F,
+) -> Range<usize> {
+    // walk backwards and discover
+    let start = chars[..idx]
+        .iter()
+        .rev()
+        .take_while(|&&(_, c)| predicate(c))
+        .last()
+        .map_or(chars[idx].0, |&(i, _)| i);
+    // walk forwards
+    let end = chars[(idx + 1)..]
+        .iter()
+        .take_while(|&&(_, c)| predicate(c))
+        .last()
+        .map_or(chars[idx].0 + chars[idx].1.len_utf8(), |&(i, c)| {
+            i + c.len_utf8()
+        });
+
+    start..end
+}
+
+fn choose_category_range<R: Rand>(
+    rand: &mut R,
+    string: &str,
+) -> Option<(Range<usize>, &'static [(u32, u32)])> {
+    let chars = string.char_indices().collect::<Vec<_>>();
+    let idx = rand.below(chars.len() as u64) as usize;
+    let c = chars[idx].1;
+
+    // figure out the categories for this char
+    let expanded = c as u32;
+    #[cfg(test)]
+    let mut names = Vec::new();
+    let mut categories = Vec::new();
+    for (_name, category) in unicode_categories::BY_NAME.iter() {
+        if get_subcategory(expanded, category).is_some() {
+            #[cfg(test)]
+            names.push(_name);
+            categories.push(category);
+        }
+    }
+
+    // ok -- we want to bias towards smaller regions to keep the mutations "tight" to original
+    // we sort the options by descending length, then pick isqrt of below(n^2)
+
+    categories.sort_by_cached_key(|cat| {
+        usize::MAX
+            - cat
+                .iter()
+                .map(|&(min, max)| (max - min + 1) as usize)
+                .sum::<usize>()
+    });
+    let options = categories.len() * categories.len();
+    let selected_idx = libafl_bolts::math::integer_sqrt(rand.below(options as u64)) as usize;
+
+    let selected = categories[selected_idx];
+
+    #[cfg(test)]
+    println!("category for `{c}' ({}): {}", c as u32, names[selected_idx]);
+
+    Some((
+        find_range(&chars, idx, |c| {
+            get_subcategory(c as u32, selected).is_some()
+        }),
+        selected,
+    ))
+}
+
+fn choose_subcategory_range<R: Rand>(
+    rand: &mut R,
+    string: &str,
+) -> Option<(Range<usize>, (u32, u32))> {
+    let chars = string.char_indices().collect::<Vec<_>>();
+    let idx = rand.below(chars.len() as u64) as usize;
+    let c = chars[idx].1;
+
+    // figure out the categories for this char
+    let expanded = c as u32;
+    #[cfg(test)]
+    let mut names = Vec::new();
+    let mut subcategories = Vec::new();
+    for (_name, category) in unicode_categories::BY_NAME.iter() {
+        if let Some(subcategory) = get_subcategory(expanded, category) {
+            #[cfg(test)]
+            names.push(_name);
+            subcategories.push(subcategory);
+        }
+    }
+
+    // see reasoning for selection pattern in choose_category_range
+
+    subcategories.sort_by_key(|&(min, max)| max - min + 1);
```

Comment:
we are favoring longer subcategories in `choose_subcategory_range`
but we are doing the opposite by favoring shorter in `choose_category_range`? (because in that function we subtract length from u32::MAX)
is this intended?

Reply:
Nope! good catch

Reply:
See updated.